File: solutions/0054-spiral-matrix/solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ans = []
        i,j = 0,0
        right, down, left, up = 0, 1, 2, 3
        row = len(matrix)
        col = len(matrix[0])
        right_wall, down_wall, left_wall, up_wall = col, row, -1, 0

        direction = right
        
        while len(ans) != row*col:
            if direction == right:
                while j < right_wall:
                    ans.append(matrix[i][j])
                    j += 1
                direction = down
                i += 1
                j -= 1
                right_wall -= 1
            elif direction == down:
                while i < down_wall:
                    ans.append(matrix[i][j])
                    i += 1
                direction = left
                i -= 1
                j -= 1
                down_wall -= 1
            elif direction == left:
                while j > left_wall:
                    ans.append(matrix[i][j])
                    j -= 1
                direction = up
                i -= 1
                j += 1
                left_wall += 1
            else:
                while i > up_wall:
                    ans.append(matrix[i][j])
                    i -= 1
                direction = right
                i += 1
                j += 1
                up_wall += 1
        
        return ans
                

File: solutions/0054-spiral-matrix/solution.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int right = 0, down = 1, left = 2, up = 3;
        int direction;

        int row = matrix.length;
        int col = matrix[0].length;

        int right_wall = col, down_wall = row, left_wall = -1, up_wall = 0; 
        
        List<Integer> ans = new ArrayList<>();

        direction = right;

        int i = 0, j = 0;

        while(ans.size() < row*col)
        {
            if (direction == right)
            {
                while(j < right_wall)
                {
                    ans.add(matrix[i][j]);
                    j++;
                }
                direction = down;
                i++;
                j--;
                right_wall--;
            }
            else if(direction == down)
            {
                while(i < down_wall)
                {
                    ans.add(matrix[i][j]);
                    i++;
                }
                direction = left;
                i--;
                j--;
                down_wall--;
            }
            else if(direction == left)
            {
                while(j > left_wall)
                {
                    ans.add(matrix[i][j]);
                    j--;
                }
                direction = up;
                i--;
                j++;
                left_wall++;
            }
            else
            {
                while(i > up_wall)
                {
                    ans.add(matrix[i][j]);
                    i--;
                }
                direction = right;
                i++;
                j++;
                up_wall++;
            }
        }

        return ans;
    }
}
